File: celesta/celesta-test/score/navigation/testNavigation.py
# coding=UTF-8

from navigation._navigation_orm import navigationTableCursor
from ru.curs.celesta.unit import TestClass, CelestaTestCase

@TestClass
class TestNavigation(CelestaTestCase):

    def testSimpleNext(self):
        c = navigationTableCursor(self.context)
        self._prepareTableForTest(c)

        c.orderBy('numb')
        c.first()
        self.assertEquals(1, c.numb)
        c.next()
        self.assertEquals(2, c.numb)
        c.next()
        self.assertEquals(3, c.numb)
        c.next()
        self.assertEquals(4, c.numb)
        c.next()
        self.assertEquals(5, c.numb)
        c.next()
        self.assertEquals(5, c.numb)


    def testSimplePrevious(self):
        c = navigationTableCursor(self.context)
        self._prepareTableForTest(c)

        c.orderBy('numb')
        c.last()
        self.assertEquals(5, c.numb)
        c.previous()
        self.assertEquals(4, c.numb)
        c.previous()
        self.assertEquals(3, c.numb)
        c.previous()
        self.assertEquals(2, c.numb)
        c.previous()
        self.assertEquals(1, c.numb)
        c.previous()
        self.assertEquals(1, c.numb)


    def testNavigateWithOffset(self):
        c = navigationTableCursor(self.context)
        self._prepareTableForTest(c)

        c.orderBy('numb')
        c.first()
        self.assertEquals(1, c.numb)

        c.navigate('>', 3)
        self.assertEquals(4, c.numb)

        c.navigate('<', 2)
        self.assertEquals(2, c.numb)

        self.assertFalse(c.navigate('<', 10))
        self.assertEquals(2, c.numb)

        self.assertFalse(c.navigate('>', 10))
        self.assertEquals(2, c.numb)


    def _prepareTableForTest(self, c):
        c.deleteAll()
        self._insert(c, 1)
        self._insert(c, 2)
        self._insert(c, 3)
        self._insert(c, 4)
        self._insert(c, 5)
        c.clear()


    def _insert(self, c, numb):
        c.clear()
        c.numb = numb
        c.insert()
        c.clear()

File: celesta-test/src/test/java/ru/curs/celesta/script/TestNavigation.java
package ru.curs.celesta.script;

import navigation.NavigationTableCursor;
import org.junit.jupiter.api.TestTemplate;
import ru.curs.celesta.CallContext;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;

import ru.curs.celesta.CelestaException;

class TestNavigation implements ScriptTest {

    @TestTemplate
    void testSimpleNext(CallContext context) {
        NavigationTableCursor c = new NavigationTableCursor(context);
        _prepareTableForTest(c);

        c.orderBy(NavigationTableCursor.numb_COLUMN);
        c.first();
        assertEquals(1, c.getNumb().intValue());
        c.next();
        assertEquals(2, c.getNumb().intValue());
        c.next();
        assertEquals(3, c.getNumb().intValue());
        c.next();
        assertEquals(4, c.getNumb().intValue());
        c.next();
        assertEquals(5, c.getNumb().intValue());
        c.next();
        assertEquals(5, c.getNumb().intValue());


    }

    @TestTemplate
    void testSimplePrevious(CallContext context) {
        NavigationTableCursor c = new NavigationTableCursor(context);
        _prepareTableForTest(c);

        c.orderBy(NavigationTableCursor.numb_COLUMN);
        c.last();
        assertEquals(5, c.getNumb().intValue());
        c.previous();
        assertEquals(4, c.getNumb().intValue());
        c.previous();
        assertEquals(3, c.getNumb().intValue());
        c.previous();
        assertEquals(2, c.getNumb().intValue());
        c.previous();
        assertEquals(1, c.getNumb().intValue());
        c.previous();
        assertEquals(1, c.getNumb().intValue());


    }

    @TestTemplate
    void testNavigateWithOffset(CallContext context) {
        NavigationTableCursor c = new NavigationTableCursor(context);
        _prepareTableForTest(c);

        c.orderBy(NavigationTableCursor.numb_COLUMN);
        c.first();
        assertEquals(1, c.getNumb().intValue());

        c.navigate(">", 3);
        assertEquals(4, c.getNumb().intValue());

        c.navigate("<", 2);
        assertEquals(2, c.getNumb().intValue());

        assertFalse(c.navigate("<", 10));
        assertEquals(2, c.getNumb().intValue());

        assertFalse(c.navigate(">", 10));
        assertEquals(2, c.getNumb().intValue());

    }

    @TestTemplate
    void testCelestaExceptionWhenOffsetLessThanZero(CallContext context) {
        NavigationTableCursor c = new NavigationTableCursor(context);
        _prepareTableForTest(c);

        c.orderBy(NavigationTableCursor.numb_COLUMN);
        c.first();
        assertEquals(1, c.getNumb().intValue());

        assertThrows(CelestaException.class,
                () -> c.navigate(">", -1));


    }

    void _prepareTableForTest(NavigationTableCursor c) {
        c.deleteAll();
        _insert(c, 1);
        _insert(c, 2);
        _insert(c, 3);
        _insert(c, 4);
        _insert(c, 5);
        c.clear();


    }

    void _insert(NavigationTableCursor c, int numb) {
        c.clear();
        c.setNumb(numb);
        c.insert();
        c.clear();
    }
}
